crates/refunder/src/submitter.rs

```diff
@@ -143,37 +120,52 @@ impl Submitter {
 }
 
 fn calculate_submission_gas_price(
-    gas_price_of_last_submission: Option<U256>,
+    gas_price_of_last_submission: Option<GasPrice>,
     web3_gas_estimation: GasPrice1559,
     newest_nonce: U256,
     nonce_of_last_submission: Option<U256>,
 ) -> Result<GasPrice> {
     let GasPrice1559 {
         max_fee_per_gas, ..
     }: gas_estimation::GasPrice1559 = web3_gas_estimation;
+
     // The gas price of the refund tx is the current prevailing gas price
     // of the web3 gas estimation plus a buffer.
     // Since we are using Eip1559 gas specification,
     // we will only pay the buffer if it is used
     let mut new_max_fee_per_gas = max_fee_per_gas * GAS_PRICE_BUFFER_FACTOR;
+    let mut new_priority_fee_per_gas = START_PRIORITY_FEE_TIP as f64;
+
     // If tx from the previous submission was not mined,
-    // we incease the gas price
+    // we incease the tip and max_gas_fee for miners
+    // in order to avoid "tx underpriced errors"
     if Some(newest_nonce) == nonce_of_last_submission {
-        // Increase the gas price to last submission's
-        // gas price plus an additional 12 %
-        if let Some(previous_gas_price) = gas_price_of_last_submission {
+        if let Some(GasPrice::Eip1559 {
+            max_fee_per_gas: max_fee_per_gas_of_last_tx,
+            max_priority_fee_per_gas: max_priority_fee_per_gas_of_last_tx,
+        }) = gas_price_of_last_submission
+        {
             new_max_fee_per_gas = f64::max(
                 new_max_fee_per_gas,
-                previous_gas_price.to_f64_lossy() * GAS_PRICE_BUMP,
+                max_fee_per_gas_of_last_tx.to_f64_lossy() * GAS_PRICE_BUMP,
             );
+            new_priority_fee_per_gas =
+                max_priority_fee_per_gas_of_last_tx.to_f64_lossy() * GAS_PRICE_BUMP;
         }
     }
-    let max_fee_per_gas = U256::from_f64_lossy(new_max_fee_per_gas);
+    if new_max_fee_per_gas > MAX_GAS_PRICE as f64 {
+        tracing::warn!(
+            "Refunding txs are likely not mined in time, as the current gas price {:?} \
+                            is higher than MAX_GAS_PRICE specified {:?}",
+            new_max_fee_per_gas,
+            MAX_GAS_PRICE
+        );
+        new_max_fee_per_gas = f64::min(MAX_GAS_PRICE as f64, new_max_fee_per_gas);
+    }
+    new_priority_fee_per_gas = f64::min(new_priority_fee_per_gas, new_max_fee_per_gas);
     let gas_price = GasPrice::Eip1559 {
-        max_fee_per_gas,
-        // The max_priority_fee_per_gas should normally allways be equal to MAX_PRIORITY_FEE_TIP
-        // Just in case the max_fee_per_gas is even lower than the constant, we need to adapt it
-        max_priority_fee_per_gas: U256::min(max_fee_per_gas, U256::from(MAX_PRIORITY_FEE_TIP)),
+        max_fee_per_gas: U256::from_f64_lossy(new_max_fee_per_gas),
+        max_priority_fee_per_gas: U256::from_f64_lossy(new_priority_fee_per_gas),
     };
     Ok(gas_price)
```

Comment:
```suggestion
    fn calculate_submission_gas_price(
    gas_price_of_last_submission: Option<GasPrice1559>,
    web3_gas_estimation: GasPrice1559,
    newest_nonce: U256,
    nonce_of_last_submission: Option<U256>,
) -> Result<GasPrice1559> {
    let mut new_gas_price = web3_gas_estimation.bump(GAS_PRICE_BUFFER_FACTOR);
    new_gas_price.max_priority_fee_per_gas =
        (START_PRIORITY_FEE_TIP as f64).min(new_gas_price.max_fee_per_gas); // limit to max fee

    // If tx from the previous submission was not mined,
    // we incease the tip and max_gas_fee for miners
    // in order to avoid "tx underpriced errors"
    if Some(newest_nonce) == nonce_of_last_submission {
        let gas_price_of_last_submission = gas_price_of_last_submission
            .map(|gas_price| gas_price.bump(GAS_PRICE_BUMP))
            .unwrap_or_default();

        new_gas_price.max_fee_per_gas = new_gas_price
            .max_fee_per_gas
            .max(gas_price_of_last_submission.max_fee_per_gas);
        new_gas_price.max_priority_fee_per_gas = new_gas_price
            .max_priority_fee_per_gas
            .max(gas_price_of_last_submission.max_priority_fee_per_gas);
    }

    if new_gas_price.max_fee_per_gas > MAX_GAS_PRICE as f64 {
        tracing::warn!(
            "Refunding txs are likely not mined in time, as the current gas price {:?} \
                            is higher than MAX_GAS_PRICE specified {:?}",
            new_gas_price.max_fee_per_gas,
            MAX_GAS_PRICE
        );
        new_gas_price.limit_cap(MAX_GAS_PRICE as f64);
    }

    Ok(new_gas_price)
}
```

Reply:
nit: How about `calculate_submission_gas_price` working with `GasPrice1559` only. Also `gas_parameters_of_last_tx` could be changed to `Option<GasPrice1559>`. I think this struct has a way better interface than `ethcontract::GasPrice`.

Then, you would just convert into `ethcontract::GasPrice` once you need to send the transaction, using `into_gas_price` function, but you would probably need to move it to `shared` crate instead of `solver` crate where it is now.

Reply:
to be honest, I don't like GasPrice1559 as I don't need `base_fee_per_gas`. but seeing how nicely this lib can be used to simplify the code, it was a must-go. Thanks for the re-write suggestions!